### PR TITLE
OpenShift docs

### PR DESCRIPTION
**Pull request by hierynomus**

### CODE REVIEW

File: setup/installation/openshift_install.md [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+description: install StackState on OpenShift
+---
+
+# Install StackState
+
+## Before you start
+
+Before you start the installation of StackState:
+
+* Check that your OpenShift environment meets the [requirements](../../requirements.md)
```
Comment (rb3ckers) [inline setup/installation/openshift_install.md:11]: Link is broken (I assume you attempt to link to the k8s requirements page)

File: setup/installation/openshift_install.md [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+| Username and password\*\* | `-u` `-p` | The username and password used by StackState to pull images from quay.io/stackstate repositories. |
+| License key | `-l` | The StackState license key. |
+| Admin API password | `-a` | The password for the admin API. Note that this API contains system maintenance functionality and should only be accessible by the maintainers of the StackState installation. This can be omitted from the command line, the script will prompt for it. |
+| Default password | `-d` | The password for the default user \(`admin`\) to access StackState's UI. This can be omitted from the command line, the script will prompt for it. |
+| Kubernetes cluster name | `-k` | *NOTE: For OpenShift do not enable this* |
+
+The generated file is suitable for a production setup \(i.e. redundant storage services\). It is also possible to create smaller deployments for test setups, see [development setup](development_setup.md).
```
Comment (rb3ckers) [inline setup/installation/openshift_install.md:75]: Link needs to refer to the sub dir `kubernetes_installation`

File: setup/installation/openshift_install.md [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+    securityContext:
+      enabled: false
+zookeeper:
+  securityContext:
+    enabled: false
+cluster-agent:
+  agent:
+    scc:
+      enabled: true
+  enabled: true
```
Comment (rb3ckers) [inline setup/installation/openshift_install.md:149]: This enables the agent, it is not needed for Openshift itself.

File: setup/installation/openshift_install.md [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+  stackstate:
+    cluster:
```
Comment (rb3ckers) [inline setup/installation/openshift_install.md:151]: Neither is the clustername and auth token

File: setup/installation/openshift_install.md [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+      name: <CLUSTER_NAME>
+      authToken: <RANDOM_TOKEN>
+  kube-state-metrics:
+    podAnnotations:
+      ad.stackstate.com/kube-state-metrics.check_names: '["kubernetes_state"]'
+      ad.stackstate.com/kube-state-metrics.init_configs: '[{}]'
+      ad.stackstate.com/kube-state-metrics.instances: '[{"kube_state_url":"http://%%host%%:%%port%%/metrics","labels_mapper":{"namespace":"kube_namespace","label_deploymentconfig":"oshift_deployment_config","label_deployment":"oshift_deployment"},"label_joins":{"kube_pod_labels":{"label_to_match":"pod","labels_to_get":["label_deployment","label_deploymentconfig"]}}}]'
```
Comment (rb3ckers) [inline setup/installation/openshift_install.md:158]: Or these annotation

File: setup/installation/openshift_install.md [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+    all:
+      securityContext:
+        enabled: false
+    kafkaTopicCreate:
+      securityContext:
+        enabled: false
+    ui:
+      securityContext:
+        enabled: false
+  stackpacks:
```
Comment (rb3ckers) [inline setup/installation/openshift_install.md:107]: This is not really needed I think.

Comment (rb3ckers) [reply setup/installation/openshift_install.md:107]: Maybe we can put the agent stuff in a separate example? OTOH it is probably easier to just tell them to install the Openshift stackpack and follow the directions to install the agent from there.

File: setup/installation/openshift_install.md [hunk 7]
```diff
@@ -0,0 +1,319 @@
...
+kubectl port-forward service/stackstate-router 8080:8080 --namespace stackstate
+```
+
+StackState will now be available in your browser at `https://localhost:8080`. Log in with the username `admin` and the default password provided in the `values.yaml` file.
+
+Next steps are
+
+* Configure [ingress](ingress.md)
+* Install a [StackPack](../../../stackpacks/about-stackpacks.md) or two
+* Give your [co-workers access](../../authentication.md).
```
Comment (rb3ckers) [inline setup/installation/openshift_install.md:215]: First and last link are broken

File: setup/installation/openshift_install.md [hunk 8]
```diff
@@ -0,0 +1,319 @@
...
+- configMap
+- downwardAPI
+- emptyDir
+- persistentVolumeClaim
+- projected
+- secret
+```
+{% endtab %}
+
+{% tab title="agent-scc.yaml" %}
```
Comment (rb3ckers) [inline setup/installation/openshift_install.md:262]: I wonder if we should even bother with the agent installation here at all. This is (or should be) already part of the agent installation itself and as such should be covered by the Openshift StackPack installation.

File: setup/installation/openshift_install.md [hunk 9]
```diff
@@ -0,0 +1,319 @@
...
+The script requires the following configuration options:
+
+| Configuration | Flag | Description |
+| :--- | :--- | :--- |
+| Base URL | `-b` | The external URL for StackState that users and agents will use to connect. For example `https://stackstate.internal`.  If you haven't decided on an Ingress configuration yet, use `http://localhost:8080`. This can be updated later in the generated file. |
+| Username and password\*\* | `-u` `-p` | The username and password used by StackState to pull images from quay.io/stackstate repositories. |
+| License key | `-l` | The StackState license key. |
+| Admin API password | `-a` | The password for the admin API. Note that this API contains system maintenance functionality and should only be accessible by the maintainers of the StackState installation. This can be omitted from the command line, the script will prompt for it. |
+| Default password | `-d` | The password for the default user \(`admin`\) to access StackState's UI. This can be omitted from the command line, the script will prompt for it. |
+| Kubernetes cluster name | `-k` | *NOTE: For OpenShift do not enable this* |
```
Comment (rb3ckers) [inline setup/installation/openshift_install.md:73]: I would change this
> *NOTE: For OpenShift do not enable this*

to include 
> instead install the Openshift stackpack and follow its  agent installation instructions via the StackState UI

File: setup/installation/openshift_install.md [hunk 10]
```diff
@@ -0,0 +1,319 @@
...
+```text
+> oc adm policy add-scc-to-user stackstate-agent-scc system:serviceaccount:stackstate:stackstate-cluster-agent-agent
+> oc adm policy add-scc-to-user stackstate-hdfs-scc system:serviceaccount:stackstate:stackstate-hbase-hdfs
+```
+
+Once this is done, you can continue with the installation of the StackState Helm chart on OpenShift.
+
+## See also
+
+- [Manage a StackState Kubernetes installation](/setup/installation/kubernetes_install/README.md)
```
Comment (rb3ckers) [inline setup/installation/openshift_install.md:319]: I would add a sentence here to clarify that for all other configuration / setup the user should refer to the Kuberentes docs.

File: setup/installation/openshift_install.md (hunk 1 above)
Comment (hierynomus) [reply setup/installation/openshift_install.md:11]: DOne!

File: setup/installation/openshift_install.md (hunk 2 above)
Comment (hierynomus) [reply setup/installation/openshift_install.md:75]: Done

File: setup/installation/openshift_install.md (hunk 3 above)
Comment (hierynomus) [reply setup/installation/openshift_install.md:149]: Moved to new chapter

File: setup/installation/openshift_install.md (hunk 4 above)
Comment (hierynomus) [reply setup/installation/openshift_install.md:151]: Moved

File: setup/installation/openshift_install.md (hunk 5 above)
Comment (hierynomus) [reply setup/installation/openshift_install.md:158]: Moved

File: setup/installation/openshift_install.md (hunk 6 above)
Comment (hierynomus) [reply setup/installation/openshift_install.md:107]: Moved to new chapter

File: setup/installation/openshift_install.md (hunk 7 above)
Comment (hierynomus) [reply setup/installation/openshift_install.md:215]: Updated

File: setup/installation/openshift_install.md (hunk 8 above)
Comment (hierynomus) [reply setup/installation/openshift_install.md:262]: Kept as discussed

File: setup/installation/openshift_install.md (hunk 9 above)
Comment (hierynomus) [reply setup/installation/openshift_install.md:73]: Pointing to new chapter now

File: setup/installation/openshift_install.md (hunk 10 above)
Comment (hierynomus) [reply setup/installation/openshift_install.md:319]: Added

File: setup/installation/openshift_install.md [hunk 11]
```diff
@@ -0,0 +1,342 @@
...
+
+# Install StackState
+
+## Before you start
+
+Before you start the installation of StackState:
+
+* Check that your OpenShift environment meets the [requirements](../requirements.md)
+* Request access credentials to pull the StackState Docker images from [StackState support](https://support.stackstate.com/).
+* Ensure you have the OpenShift command line tools installed (`oc`)
```
Comment (gggina) [inline setup/installation/openshift_install.md:13]: Are there instructions to install this? do you use `oc` to check it is installed?

File: setup/installation/openshift_install.md [hunk 12]
```diff
@@ -0,0 +1,342 @@
...
+### Create project
+
+Start by creating the project where you want to install StackState. In our walkthrough we will use the namespace `stackstate`:
+
+```text
+oc new-project stackstate
+```
+
+{% hint style="info" %}
+For `helm` and `kubectl` commands the project name is used as the namespace name in the `--namespace` flag
```
Comment (gggina) [inline setup/installation/openshift_install.md:39]: The project name is used in `helm` and `kubectl` commands as the namespace name in the `--namespace` flag.

File: setup/installation/openshift_install.md [hunk 13]
```diff
@@ -0,0 +1,342 @@
...
+
+| Configuration | Flag | Description |
+| :--- | :--- | :--- |
+| Base URL | `-b` | The external URL for StackState that users and agents will use to connect. For example `https://stackstate.internal`.  If you haven't decided on an Ingress configuration yet, use `http://localhost:8080`. This can be updated later in the generated file. |
+| Username and password\*\* | `-u` `-p` | The username and password used by StackState to pull images from quay.io/stackstate repositories. |
+| License key | `-l` | The StackState license key. |
+| Admin API password | `-a` | The password for the admin API. Note that this API contains system maintenance functionality and should only be accessible by the maintainers of the StackState installation. This can be omitted from the command line, the script will prompt for it. |
+| Default password | `-d` | The password for the default user \(`admin`\) to access StackState's UI. This can be omitted from the command line, the script will prompt for it. |
+| Kubernetes cluster name | `-k` | *NOTE: For OpenShift please follow the [instructions here](openshift_install.md#automatically-installing-the-cluster-agent-for-openshift)* |
```
Comment (gggina) [inline setup/installation/openshift_install.md:73]: Would it make more sense to include those instructions before the generate values.yaml section?

File: setup/installation/openshift_install.md [hunk 14]
```diff
@@ -0,0 +1,342 @@
...
+
+The generated file is suitable for a production setup \(i.e. redundant storage services\). It is also possible to create smaller deployments for test setups, see [development setup](kubernetes_install/development_setup.md).
+
+{% hint style="info" %}
+Store the `values.yaml` file somewhere safe. You can reuse this file for upgrades, which will save time and \(more importantly\) will ensure that StackState continues to use the same API key. This is desirable as it means agents and other data providers for StackState will not need to be updated.
+{% endhint %}
+
+### Additional OpenShift values file
+
+Because OpenShift has a more strict security model than plain Kubernetes, all of the standard security contexts in the deployment need to be disabled.
```
Comment (gggina) [inline setup/installation/openshift_install.md:83]: a stricter security model

File: setup/installation/openshift_install.md [hunk 15]
```diff
@@ -0,0 +1,342 @@
...
+Furthermore, the StackState installation needs to add specific OpenShift `SecurityContextConfiguration` objects to the OpenShift installation. If you're installing using an administrator account it is possible to automatically create these as needed.
```
Comment (gggina) [inline setup/installation/openshift_install.md:84]: (new line)
If you're installing using an administrator account, it is possible

File: setup/installation/openshift_install.md [hunk 16]
```diff
@@ -0,0 +1,342 @@
...
+
+| Pod(s) | Config key | Description |
+| :--- | :--- | :--- |
+| The HBase HDFS namenodes | `hbase.hdfs.scc.enabled` | These processes need to be able to run a `chmod` on their volumes and run with a specific (pre-known) UID, in order for the clients to write to the store. |
```
Comment (gggina) [inline setup/installation/openshift_install.md:88]: For the clients to write to the store, these processes require permissions to run `chmod` on their volumes and run with a specific (pre-known) UID.

File: setup/installation/openshift_install.md [hunk 17]
```diff
@@ -0,0 +1,342 @@
...
+
+If you're _not_ using an administrator account, Please follow the [instructions below](openshift_install.md#manually-create-securitycontextconfiguration-objects) to first install the `SecurityContextConfiguration` objects in OpenShift. After that install the StackState Helm chart with the above flags set to `false`.
```
Comment (gggina) [inline setup/installation/openshift_install.md:90]: If you're _not_ using an administrator account, Please follow the instructions below to [first install the `SecurityContextConfiguration` objects in OpenShift](openshift_install.md#manually-create-securitycontextconfiguration-objects) .

File: setup/installation/openshift_install.md [hunk 18]
```diff
@@ -0,0 +1,342 @@
...
+      enabled: false
+  tephra:
+    securityContext:
+      enabled: false
+zookeeper:
+  securityContext:
+    enabled: false
+```
+
+Store this file next to the generated `values.yaml` file and name it `openshift-values.yaml`
```
Comment (gggina) [inline setup/installation/openshift_install.md:141]: name it `openshift-values.yaml`.

File: setup/installation/openshift_install.md [hunk 19]
```diff
@@ -0,0 +1,342 @@
...
+
+### Automatically installing the Cluster Agent for OpenShift
```
Comment (gggina) [inline setup/installation/openshift_install.md:143]: ### Automatically install the Cluster Agent for OpenShift

File: setup/installation/openshift_install.md [hunk 20]
```diff
@@ -0,0 +1,342 @@
...
+
+StackState has built-in support for OpenShift by means of the [OpenShift StackPack](../../stackpacks/integrations/openshift.md). To get started quickly, the StackState installation can automate installation of this StackPack and the required agent for the cluster that StackState itself will be installed on. This is not required and can always be done later from the StackPacks page of the StackState UI for StackState's cluster or any other OpenShift cluster.
```
Comment (gggina) [inline setup/installation/openshift_install.md:145]: the required Agent

File: setup/installation/openshift_install.md [hunk 21]
```diff
@@ -0,0 +1,342 @@
...
+
+The only required information is a name for the OpenShift cluster that will distinguish it from the other OpenShift clusters monitored by StackState. A good choice usually is the same name that is used in the kube context configuration. This will then automatically install the StackPack and install a Daemonset for the agent and a deployment for the so called cluster agent. For the full details, please read the [OpenShift StackPack](../../stackpacks/integrations/openshift.md) page.
+
+In order to automate this installation, the below values file can be added to the `helm install` command. The agent chart also needs an additional `SecurityContextConfiguration` on OpenShift. If you're installing as an administrator on the OpenShift cluster, it is possible to automatically create this. You can configure this using the following configuration option in the values file:
```
Comment (gggina) [inline setup/installation/openshift_install.md:149]: To automate this installation,
...
(new line)
If you're installing as an administrator on the OpenShift cluster,

File: setup/installation/openshift_install.md [hunk 22]
```diff
@@ -0,0 +1,342 @@
...
+
+| Pod(s) | Config key | Description |
+| :--- | :--- | :--- |
+| The Agent that runs the Kubernetes checks | `cluster-agent.agent.scc.enabled` | This process needs to run a privileged container with direct access to the host(network) and volumes. |
+
+If you're not installing as an administrator, please follow the [instructions below](openshift_install.md#manually-create-securitycontextconfiguration-objects) to first install the `SecurityContextConfiguration` objects in OpenShift. Then install the Helm chart with the flag set to `false`.
```
Comment (gggina) [inline setup/installation/openshift_install.md:155]: If you're not installing as an administrator, follow the instructions below to [first install the `SecurityContextConfiguration` objects in OpenShift](openshift_install.md#manually-create-securitycontextconfiguration-objects). Then install the Helm chart with the flag set to `false`.

Comment (gggina) [reply setup/installation/openshift_install.md:155]: "Then install the Helm chart with the flag set to false" - which flag?

File: setup/installation/openshift_install.md [hunk 23]
```diff
@@ -0,0 +1,342 @@
...
+Two placeholders in this file need to be given a value before this can be applied to the Helm installation:
+
+- `<CLUSTER_NAME>`: A name that StackState will use to identify the cluster
+- `<RANDOM_TOKEN>`: A 32 character random token. This can be generated by executing `head -c32 < /dev/urandom | md5sum  | cut -c-32`
+
+Save this as `agent-values.yaml` and add it to the `helm install` command to enable this feature.
+
+### Deploy StackState with Helm
+
+Use the generated `values.yaml` and copied `openshift-values.yaml` (and `agent-values.yaml`) files to deploy the latest StackState version to the `stackstate` namespace with the following command:
```
Comment (gggina) [inline setup/installation/openshift_install.md:193]: Use the generated `values.yaml` file and copied `openshift-values.yaml` file to deploy the latest StackState version to the `stackstate` namespace with the command below. if you want to automatically install the Cluster Agent for Openshift, you will also require the `agent-values.yaml`created in the previous step:

File: setup/installation/openshift_install.md [hunk 24]
```diff
@@ -0,0 +1,342 @@
...
+helm list --namespace stackstate
+
+# Check pods are running
+# It may take some time for all pods to be installed or available
+kubectl get pods --namespace stackstate
+```
+
+### Access the StackState UI
+
+After StackState has been deployed you can check if all pods are up and running:
```
Comment (gggina) [inline setup/installation/openshift_install.md:218]: After StackState has been deployed,

File: setup/installation/openshift_install.md [hunk 25]
```diff
@@ -0,0 +1,342 @@
...
+Next steps are
+
+* Configure [ingress](kubernetes_install/ingress.md)
+* Install a [StackPack](../../stackpacks/about-stackpacks.md) or two
+* Give your [co-workers access](../authentication.md).
+
+
+## Manually create `SecurityContextConfiguration` objects
+
+If you cannot use an administrative account to install StackState on OpenShift, ask your administrator to apply the below `SecurityContextConfiguration` objects.
```
Comment (gggina) [inline setup/installation/openshift_install.md:241]: If you cannot use an administrator account

File: setup/installation/openshift_install.md [hunk 26]
```diff
@@ -0,0 +1,342 @@
...
+
+{% tabs %}
```
Comment (gggina) [inline setup/installation/openshift_install.md:243]: Are both of the tabs always required? If so, I would list them one after the other? (tabs are useful for either/or content)

File: setup/installation/openshift_install.md [hunk 27]
```diff
@@ -0,0 +1,342 @@
...
+  - configMap
+  - downwardAPI
+  - emptyDir
+  - hostPath
+  - secret
+```
+{% endtab %}
+{% endtabs %}
+
+After these are applied, as administrator, execute the following commands to grant the service accounts access to these `SecurityContextConfiguration`s:
```
Comment (gggina) [inline setup/installation/openshift_install.md:329]: The last "s" looks strange.

File: setup/installation/openshift_install.md (hunk 11 above)
Comment (hierynomus) [reply setup/installation/openshift_install.md:13]: Typically you would have them already if you're running OpenShift. It's like you need to have `kubectl` when you're running Kubernetes, it's part of the package.

File: setup/installation/openshift_install.md (hunk 13 above)
Comment (hierynomus) [reply setup/installation/openshift_install.md:73]: This instruction is purely for the `-k` configuration switch, the other switches are valid.

**Review by gggina — APPROVED** [review]

**Pull request merged**